luarules/gadgets/unit_shield_behaviour.lua

```diff
@@ -0,0 +1,350 @@
+
+		shieldData.shieldCoverageChecked = true
+	end
+end
+
+local function shieldNegatesDamageCheck(unitID, unitTeam, attackerID, attackerTeam)
+	if shieldedUnits[unitID] and allTeams[unitTeam] ~= allTeams[attackerTeam] and next(shieldedUnits[unitID]) then
+		if not shieldedUnits[attackerID] or (not next(shieldedUnits[attackerID]) and next(shieldedUnits[unitID])) then
+			return true
+		end
+
+		for subKey in pairs(shieldedUnits[unitID]) do
+			if shieldedUnits[attackerID][subKey] then
+				break
+			else
+				--The units have to share all of the same shield spaces. As soon as a mismatch is found, that means they don't occupy the same shield space and the shot should be blocked.
+				return true
+			end
+		end
+	end
+
+	return false
+end
+
+local shieldUnitsTotalCount = 0
+local shieldUnitIndex = {}
+local shieldCheckFlags = {}
+local lastShieldCheckChunkNumber = 1
+local shieldCheckChunkSize = 10
+local shieldCheckEndIndex = 1
+
+function gadget:GameFrame(frame)
+	gameSeconds = spGetGameSeconds()
+
+	for shieldUnitID in pairs(shieldCheckFlags) do
+		local shieldData = shieldUnitsData[shieldUnitID]
+
+		if shieldData then
+			if shieldData.shieldDamage > 0 and shieldData.shieldWeaponNumber > -1 then
+				local enabledState, shieldPower = spGetUnitShieldState(shieldUnitID)
+				shieldPower = shieldPower - shieldData.shieldDamage
+
+				if shieldPower < 0 then
+					shieldPower = 0
+				end
+
+				spSetUnitShieldState(shieldUnitID, shieldData.shieldWeaponNumber, shieldPower)
+				shieldData.shieldDamage = 0
+
+				if shieldData.downtimeReset < gameSeconds and shieldPower <= 0 then
+					triggerDowntime(shieldUnitID, shieldData.shieldWeaponNumber)
+					removeCoveredUnits(shieldUnitID)
+				end
+			else
+				shieldData.shieldDamage = 0
+			end
+
+			shieldCheckFlags[shieldUnitID] = nil
+		end
+
+	end
+	for shieldUnitID, shieldData in pairs(shieldUnitsData) do
+		if frame % Game.gameSpeed == 0 then
+			if select(2, spGetUnitIsStunned(shieldUnitID)) and shieldData.downtimeReset ~= 0 then
+				shieldData.downtimeReset = shieldData.downtimeReset + 1
+			end
+		end
+
+		if frame % 10 == 0 then
+
+			if shieldData.shieldEnabled == false and shieldData.downtimeReset ~= 0 and shieldData.downtimeReset <= gameSeconds then
+				shieldData.downtimeReset = 0
+				shieldData.shieldEnabled = true
+				spSetUnitShieldRechargeDelay(shieldUnitID, shieldData.shieldWeaponNumber, 0)
+
+				-- This section is to allow slower moving projectiles already inside the shield when it comes back online to damage units within the radius.
+				local x, y, z = spGetUnitPosition(shieldUnitID)
+				local radius = shieldData.radius-25
+				local xmin = x - radius
+				local xmax = x + radius
+				local zmin = z - radius
+				local zmax = z + radius
+				local projectiles = spGetProjectilesInRectangle(xmin, zmin, xmax, zmax)
+				for _, projectileID in ipairs(projectiles) do
+					projectileShieldHitCache[projectileID] = true
+				end
+			end
+		end
+	end
+
+	if frame % 90 == 0 then
```

Comment:
Does this mean there can be an up to 3 second lag between a shield being built and the shield functioning?

Reply:
Additionally perform this on-demand for a shield in the `UnitFinished` callin?

Reply:
That would work beautifully. 
In fact, I think that this would allow us to keep the 90 frame ( or just a higher frame) slow refresh without noticable changes to performance.

After all, the coverage check is triggered to and allowed to be performed once per slow update in shieldpredamaged callin